### PR TITLE
Use the javac auto-discovery mechanism instead of the -processor flag

### LINKED ISSUE #20: Applying the checker plugin disables other annotation processors
Rather than add checker to the annotationProcessor configuration, I believe the plugin is replacing the existing package:
```.java
          compile.options.annotationProcessorPath = project.configurations.checkerFramework
```
I have my own annotation processor that is part of our multi-module build and needs to run before checker, otherwise there are missing classes. 

Also, specifying the `-processor` argument to `javac` disables the auto-discovery process, which means that my annotation processor is not found.

I have worked around this by adding my annotation processor to the checkerFramework configuration, but it would be nice to fix this.

### PR DESCRIPTION
**Pull request by kelloggm**
Using the `-processor` flag to specify checkers makes this plugin incompatible with other annotation processors which are typically auto-discovered by `javac`. Since that is the normal approach to annotation processing (used by Lombok and AutoValue, among others), this plugin should support it.

This PR changes the way that the plugin informs `javac` which checkers should be run. Previously, we passed the `-processor` flag manually, overriding any other processors that would have been auto-discovered (and any other processors the user manually specified). This PR instead generates a manifest file in the style of https://checkerframework.org/manual/#checker-auto-discovery and places the directory containing it on the annotation processor path. `javac` then finds the checkers via annotation processor discovery.

I've bumped the minor version number, because this is a backwards-incompatible change: the behavior is now different in the presence of other processors.

Fixes #20 if the other processors were auto-discovered, as is usual

Credit for the design goes to @msridar.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,6 +4,8 @@
 [![Build Status](https://travis-ci.com/kelloggm/checkerframework-gradle-plugin.svg?branch=master)](https://travis-ci.com/kelloggm/checkerframework-gradle-plugin)
 
 This plugin configures `JavaCompile` tasks to use the [Checker Framework](https://checkerframework.org) for pluggable type-checking.
+The plugin runs checkers by creating a custom manifest file containing the checkers you specify,
```
Comment (mernst) [inline README.md]: This implementation detail feels out of place on line 6 of the README.  I would cut it or leave it to later.

File: README.md [hunk 2]
```diff
@@ -245,6 +247,14 @@ the [Lombok Gradle Plugin](https://plugins.gradle.org/plugin/io.freefair.lombok)
 to delombok your source code before it is passed to the Checker Framework
 for typechecking. This plugin does not support any other use of Lombok.
 
+## Compatibility with other annotation processors
+
+The plugin should be compatible with other annotation processors, as long
```
Comment (mernst) [inline README.md]: "be compatible with" is ambiguous.  Be more specific.  You could cut the first sentence, because the second sentence starts with the important information.

File: README.md [hunk 3]
```diff
@@ -245,6 +247,14 @@ the [Lombok Gradle Plugin](https://plugins.gradle.org/plugin/io.freefair.lombok)
...
+as they are run via `javac`'s auto-discovery mechanism. If you already provide
+a `-processor` option to `javac`, you **cannot** use this plugin: you
+must also provide the checkers you wish to use by hand, because the `-processor`
```
Comment (mernst) [inline README.md]: 'by hand" is confusing (especially together with the vague "you"); it might mean on every invocation.
"explicitly" and "your build.gradle file" would be clearer and easier to read.

File: README.md [hunk 4]
```diff
@@ -245,6 +247,14 @@ the [Lombok Gradle Plugin](https://plugins.gradle.org/plugin/io.freefair.lombok)
...
+flag to `javac` disables annotation processor auto-discovery.
```
Comment (mernst) [inline README.md]: If you keep a note about the implementation via auto-discovery, this would be the place to put it.  Right now this sentence is confusing because it's not clear why it brings up auto-discovery (that's not clear even given the first sentence, which should be deleted in any event).

File: README.md [hunk 5]
```diff
@@ -250,6 +250,16 @@ the [Lombok Gradle Plugin](https://plugins.gradle.org/plugin/io.freefair.lombok)
 to delombok your source code before it is passed to the Checker Framework
 for typechecking. This plugin does not support any other use of Lombok.
 
+## Compatibility with other annotation processors
+
+The plugin runs checkers by creating a custom manifest file containing the 
+checkers you specify, and then allowing `javac`'s processor auto-discovery 
+feature to find them. If you already provide
+a `-processor` option to `javac`, you **cannot** use this plugin: you
```
Comment (mernst) [inline README.md]: This run-on sentence is confusing.  Does the text after the colon refer to your inability to use this plugin, or to how to work around the problem?
Another solution would be for the programmer to use auto-discovery.  Is there a reason that is not mentioned as an option?
Does the plugin respect existing manifest files and augment them?  (I determined the answer is "yes" by reading the code.  But removing the discussion of manifest files from the user manual means that the user manual doesn't have to even discuss this.)

File: src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy [hunk 6]
```diff
@@ -193,24 +199,34 @@ final class CheckerFrameworkPlugin implements Plugin<Project> {
...
-          compile.options.compilerArgs = [
+          compile.options.compilerArgs += [
             "-Xbootclasspath/p:${project.configurations.checkerFrameworkAnnotatedJDK.asPath}".toString()
           ]
           if (!userConfig.checkers.empty) {
-            compile.options.compilerArgs << "-processor" << userConfig.checkers.join(",")
+            compile.dependsOn(createManifestTask)
+            // Add the manifest file to the annotation processor path, so that the javac
+            // annotation processor discovery mechanism finds the checkers to use and
+            // runs them alongside any other annotation processors on the processor path.
```
Comment (mernst) [inline src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy]: This makes it sound like every annotation processor on the processor path will be run.  I don't think that is what you mean (but if it is, please clarify).  I think you actually mean any checker mentioned in a manifest file that is on the annotation processor path.

File: src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy [hunk 7]
```diff
@@ -193,24 +199,34 @@ final class CheckerFrameworkPlugin implements Plugin<Project> {
...
+            // Using the -processor flag to specify checkers would make the plugin incompatible
+            // with other annotation processors, which are typically auto-discovered by javac.
```
Comment (mernst) [inline src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy]: "typically" may be true, but it is not relevant.  Just "annotation processors that are auto-discovered" is shorter and keeps the focus where it belongs.

File: src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy [hunk 8]
```diff
@@ -193,24 +199,34 @@ final class CheckerFrameworkPlugin implements Plugin<Project> {
...
+            // Since that is the normal approach to annotation processing (used by Lombok and
```
Comment (mernst) [inline src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy]: Cut this sentence.

File: README.md [hunk 9]
```diff
@@ -250,6 +250,16 @@ the [Lombok Gradle Plugin](https://plugins.gradle.org/plugin/io.freefair.lombok)
 to delombok your source code before it is passed to the Checker Framework
 for typechecking. This plugin does not support any other use of Lombok.
 
+## Compatibility with other annotation processors
```
Comment (mernst) [inline README.md]: This section is really about incompatibility with the `-processor` command-line argument.

File: README.md [hunk 10]
```diff
@@ -250,6 +250,16 @@ the [Lombok Gradle Plugin](https://plugins.gradle.org/plugin/io.freefair.lombok)
...
+
+The plugin runs checkers by creating a custom manifest file containing the
```
Comment (mernst) [inline README.md]: This first sentence is true, but is not what the section is about, and therefore is a bad way to start it.  Begin with what the reader cares about, which is the second sentence.  Delete the first sentence, which is more appropriate for developer documentation than the user manual.

File: src/main/groovy/org/checkerframework/gradle/plugin/CreateManifestTask.groovy [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        manifestFile << this.checkers.join('\n')
+    }
+
+    /**
+     * Required so that this can incrementalized correctly, and so that `gradle clean build` behaves
+     * the same as `gradle clean ; gradle build`.
+     */
+    @OutputFile
+    def getManifestLocation() {
+        return "${project.buildDir}/checkerframework/META-INF/services/javax.annotation.processing.Processor"
```
Comment (mernst) [inline src/main/groovy/org/checkerframework/gradle/plugin/CreateManifestTask.groovy]: Minor:  there are two occurrences of the directory string, so you might consider abstracting that out.

File: README.md [hunk 12]
```diff
@@ -250,6 +250,16 @@ the [Lombok Gradle Plugin](https://plugins.gradle.org/plugin/io.freefair.lombok)
 to delombok your source code before it is passed to the Checker Framework
 for typechecking. This plugin does not support any other use of Lombok.
 
+## Incompatibility with explicitly-specified annotation processors
+
+If you provide a `-processor` option to `javac`,
+this plugin will have no effect, because it relies on
+`javac`'s annotation processor discovery mechanism. The
+`-processor` flag disables the discovery mechanism. You should
```
Comment (mernst) [inline README.md]: Move this sentence after "have no effect".  Put the explanation of what to do first, and put it together rather than breaking it up with explanation or background.

File: README.md [hunk 13]
```diff
@@ -250,6 +250,16 @@ the [Lombok Gradle Plugin](https://plugins.gradle.org/plugin/io.freefair.lombok)
...
+either replace your `-processor` flag with a manifest file
+(as described [here](https://checkerframework.org/manual/#checker-auto-discovery)),
```
Comment (mernst) [inline README.md]: I dislike "here" as anchor text, because it is not very informative.  How about dropping the parenthesis and using "manifest file" as the anchor text?

File: src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy [hunk 14]
```diff
@@ -193,24 +199,32 @@ final class CheckerFrameworkPlugin implements Plugin<Project> {
...
 
       project.tasks.withType(AbstractCompile).all { compile ->
         if (compile.hasProperty('options') && (!userConfig.excludeTests || !compile.name.toLowerCase().contains("test"))) {
           compile.options.annotationProcessorPath =
                   compile.options.annotationProcessorPath == null ?
                           project.configurations.checkerFramework :
                           project.configurations.checkerFramework.plus(compile.options.annotationProcessorPath)
-            // Check whether to use the Error Prone javac
-            if (needErrorProneJavac) {
+          // Check whether to use the Error Prone javac
```
Comment (mernst) [inline src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy]: This indentation change is good, but in general it would be better to put it in a separate commit, to improve understandability both during code review and afterward.

File: src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy [hunk 15]
```diff
@@ -193,24 +199,32 @@ final class CheckerFrameworkPlugin implements Plugin<Project> {
...
+          compile.options.compilerArgs += [
             "-Xbootclasspath/p:${project.configurations.checkerFrameworkAnnotatedJDK.asPath}".toString()
           ]
           if (!userConfig.checkers.empty) {
-            compile.options.compilerArgs << "-processor" << userConfig.checkers.join(",")
+            compile.dependsOn(createManifestTask)
+            // Add the manifest file to the annotation processor path, so that the javac
+            // annotation processor discovery mechanism finds the checkers to use and
+            // runs them alongside any other auto-discovered annotation processors.
+            // Using the -processor flag to specify checkers would make the plugin incompatible
```
Comment (mernst) [inline src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy]: because the -processor flag disables auto-discovery.

File: src/main/groovy/org/checkerframework/gradle/plugin/CreateManifestTask.groovy [hunk 16]
```diff
@@ -0,0 +1,44 @@
+package org.checkerframework.gradle.plugin
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.OutputFile
+import org.gradle.api.tasks.TaskAction
+
+/**
+ * This task generates a manifest file in the style of
+ * https://checkerframework.org/manual/#checker-auto-discovery
```
Comment (mernst) [inline src/main/groovy/org/checkerframework/gradle/plugin/CreateManifestTask.groovy]: I would add a period at the end of the sentence.  Otherwise, it's hard to read.

File: README.md [hunk 17]
```diff
@@ -250,6 +250,16 @@ the [Lombok Gradle Plugin](https://plugins.gradle.org/plugin/io.freefair.lombok)
 to delombok your source code before it is passed to the Checker Framework
 for typechecking. This plugin does not support any other use of Lombok.
 
+## Incompatibility with explicitly-specified annotation processors
+
+If you provide a `-processor` option to `javac`,
```
Comment (mernst) [inline README.md]: Can you add a run-time check for whether a `-processor` command-line argument is present?  That will be less error-prone than expecting users to pay attention to this section of the manual.

An even more user-friendly design would be to have two codepaths in the plugin:  if -processor is present, the plugin augments it; if -processor is not present, the plugin creates a manifest and javac does auto-discovery.

File: src/main/groovy/org/checkerframework/gradle/plugin/CreateManifestTask.groovy [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+ * Once the plugin places this directory containing it on the annotation processor path,
+ * javac can find the checkers via annotation processor discovery.
+ */
+class CreateManifestTask extends DefaultTask {
+
+    private final static String buildDirectory = "checkerframework/META-INF/services/"
```
Comment (mernst) [inline src/main/groovy/org/checkerframework/gradle/plugin/CreateManifestTask.groovy]: Could you add ${project.buildDir} here instead of in two places below?  Also, shouldn't this be manifestDirectory rather than buildDirectory?

File: src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy (hunk 14 above)
Comment (kelloggm) [reply src/main/groovy/org/checkerframework/gradle/plugin/CheckerFrameworkPlugin.groovy]: I agree, but it's not important enough to delay this PR over.

File: src/main/groovy/org/checkerframework/gradle/plugin/CreateManifestTask.groovy (hunk 18 above)
Comment (kelloggm) [reply src/main/groovy/org/checkerframework/gradle/plugin/CreateManifestTask.groovy]: No. These are static constants, but `${project.buildDir}` is a property of the project that will be different between instantiations of this class.

**Comment by kelloggm** [comment]
That is issue is #50
